### ISSUE TITLE
Upgrade issue 1.60 and 1.6.1 or higher

### ISSUE THREAD

**Issue by kyleabcha** [bcgit/bc-java#575]
Hi,

we have an issue when upgrading bcprov-jdk15on-160.jar and bctls-jdk15on-160.jar to bcprov-jdk15on-161.jar and bctls-jdk15on-161.jar or higher versions

> java.lang.InternalError: unable to open random source
	at org.bouncycastle.jcajce.provider.drbg.DRBG$URLSeededSecureRandom$1.run(DRBG.java:294)
	at org.bouncycastle.jcajce.provider.drbg.DRBG$URLSeededSecureRandom$1.run(DRBG.java:285)
	at java.security.AccessController.doPrivileged(Native Method)
	at org.bouncycastle.jcajce.provider.drbg.DRBG$URLSeededSecureRandom.<init>(DRBG.java:284)
	at org.bouncycastle.jcajce.provider.drbg.DRBG.createCoreSecureRandom(DRBG.java:131)
	at org.bouncycastle.jcajce.provider.drbg.DRBG.createInitialEntropySource(DRBG.java:115)
	at org.bouncycastle.jcajce.provider.drbg.DRBG.access$400(DRBG.java:29)
	at org.bouncycastle.jcajce.provider.drbg.DRBG$HybridSecureRandom.<init>(DRBG.java:357)
	at org.bouncycastle.jcajce.provider.drbg.DRBG.createBaseRandom(DRBG.java:179)
	at org.bouncycastle.jcajce.provider.drbg.DRBG.access$100(DRBG.java:29)
	at org.bouncycastle.jcajce.provider.drbg.DRBG$Default.<clinit>(DRBG.java:193)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:39)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:27)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:513)
	at java.lang.Class.newInstance0(Class.java:357)
	at java.lang.Class.newInstance(Class.java:310)
	at java.security.Provider$Service.newInstance(Provider.java:1221)
	at sun.security.jca.GetInstance.getInstance(GetInstance.java:220)
	at sun.security.jca.GetInstance.getInstance(GetInstance.java:147)
	at java.security.SecureRandom.getInstance(SecureRandom.java:254)
	at java.security.SecureRandom.getDefaultPRNG(SecureRandom.java:176)
	at java.security.SecureRandom.<init>(SecureRandom.java:133)

when enabled debug log. It is an error about \dev\urandom

> java.io.FileNotFoundException: \dev\urandom (The system cannot find the path specified)
	at java.io.FileInputStream.open(Native Method)
	at java.io.FileInputStream.<init>(FileInputStream.java:120)
	at java.io.FileInputStream.<init>(FileInputStream.java:79)
	at sun.net.www.protocol.file.FileURLConnection.connect(FileURLConnection.java:70)
	at sun.net.www.protocol.file.FileURLConnection.getInputStream(FileURLConnection.java:161)
	at java.net.URL.openStream(URL.java:1010)
	at org.bouncycastle.jcajce.provider.drbg.DRBG$URLSeededSecureRandom$1.run(DRBG.java:290)
	at org.bouncycastle.jcajce.provider.drbg.DRBG$URLSeededSecureRandom$1.run(DRBG.java:1)
	at java.security.AccessController.doPrivileged(Native Method)
	at org.bouncycastle.jcajce.provider.drbg.DRBG$URLSeededSecureRandom.<init>(DRBG.java:284)
	at org.bouncycastle.jcajce.provider.drbg.DRBG.createCoreSecureRandom(DRBG.java:131)
	at org.bouncycastle.jcajce.provider.drbg.DRBG.createInitialEntropySource(DRBG.java:115)
	at org.bouncycastle.jcajce.provider.drbg.DRBG.access$2(DRBG.java:77)
	at org.bouncycastle.jcajce.provider.drbg.DRBG$HybridSecureRandom.<init>(DRBG.java:358)
	at org.bouncycastle.jcajce.provider.drbg.DRBG.createBaseRandom(DRBG.java:179)
	at org.bouncycastle.jcajce.provider.drbg.DRBG.access$0(DRBG.java:162)
	at org.bouncycastle.jcajce.provider.drbg.DRBG$Default.<clinit>(DRBG.java:193)

There is no issue when using the 160 version.

I am using Java 6 32 bit on Windows 10 64 bit.

Could someone help me with this?

Thanks.

**Comment by bcgit** [comment]
You would only see this if there is a property called securerandom.source set and it is not pointing to anything that can be read.

**Comment by kyleabcha** [comment]
Another issue is that
```
java.lang.SecurityException: JCE cannot authenticate the provider BC
	at javax.crypto.SunJCE_b.a(DashoA13*..)
	at javax.crypto.KeyAgreement.getInstance(DashoA13*..)
	at bc.test.BCUpgradeTest.main(BCUpgradeTest.java:42)
Caused by: java.util.jar.JarException: Cannot parse file:/C:/Program%20Files%20(x86)/Java/jdk1.6.0_45/jre/lib/ext/bcprov-ext-jdk15on-162.jar
	at javax.crypto.SunJCE_c.a(DashoA13*..)
	at javax.crypto.SunJCE_b.b(DashoA13*..)
	at javax.crypto.SunJCE_b.a(DashoA13*..)
	... 3 more
```

I added ``bcprov-ext-jdk15on-162.jar``  to ``.../jre/lib/ext``

**Comment by bcgit** [comment]
What JVM is this on? It's not Oracle 1.7 or earlier is it?

**Comment by kyleabcha** [comment]
I am using Oracle Java 1.6

java version "1.6.0_45"
Java(TM) SE Runtime Environment (build 1.6.0_45-b06)
Java HotSpot(TM) Client VM (build 20.45-b01, mixed mode, sharing)

**Comment by bcgit** [comment]
Try https://downloads.bouncycastle.org/misc/bcprov-jdk15to18-1.62.jar

**Comment by kyleabcha** [comment]
the same issue as before.

> Caused by: java.util.jar.JarException: Cannot parse file:/C:/Program%20Files%20(x86)/Java/jdk1.6.0_45/jre/lib/ext/bcprov-jdk15to18-1.62.jar

**Comment by bcgit** [comment]
The jar file is corrupted. My guess would be that something has copied it in thinking it is a text file.

**Comment by kyleabcha** [comment]
hi,
I still open the jar file and see its content by using WinZip for windows

**Comment by bcgit** [comment]
Hmm... Try doing a "jar tf bcprov-jdk15to18-1.62.jar"

**Comment by kyleabcha** [comment]
it is working as below

[https://github.com/kyleabcha/repo/blob/master/bc_files.png](url)

**Comment by bcgit** [comment]
Hmm. Okay, remove the signatures in the jar and try signing it yourself. If it's the current signature that's causing the parsing error, the JCE will reject it, but you should find the signature algorithms work.

**Comment by bcgit** [comment]
Try: https://downloads.bouncycastle.org/misc/bcprov-jdk15to18-163.jar

It's signed using Oracle 7.

**Comment by kyleabcha** [comment]
thanks, this version is working fine.

**Comment by peterdettman** [comment]
Fixed in 1.63
